metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorFactoryDecorator.java

```diff
@@ -0,0 +1,105 @@
+package com.netflix.metacat.common.server.connectors;
+
+import com.netflix.metacat.common.server.api.ratelimiter.RateLimiter;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A decorator for a connector factory to add additional cross-cutting functionality
+ * to all connector services.
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class ConnectorFactoryDecorator implements ConnectorFactory {
+    private final ConnectorPlugin connectorPlugin;
+    @Getter
+    private final ConnectorFactory delegate;
+    private final ConnectorContext connectorContext;
+    private final RateLimiter rateLimiter;
+
+    @Override
+    public ConnectorCatalogService getCatalogService() {
+        ConnectorCatalogService service = delegate.getCatalogService();
+
+        if (isRateLimiterEnabled()) {
+            log.info("Creating rate-limited connector catalog services for connector-type: {}, "
+                         + "plugin-type: {}, catalog: {}, shard: {}",
+                connectorContext.getConnectorType(), connectorPlugin.getType(),
+                connectorContext.getCatalogName(), connectorContext.getCatalogShardName());
+            service = new ThrottlingConnectorCatalogService(service, rateLimiter);
+        }
+
+        return service;
+    }
+
+    @Override
+    public ConnectorDatabaseService getDatabaseService() {
+        ConnectorDatabaseService service = delegate.getDatabaseService();
+
+        if (isRateLimiterEnabled()) {
+            log.info("Creating rate-limited connector database services for connector-type: {}, "
+                         + "plugin-type: {}, catalog: {}, shard: {}",
+                connectorContext.getConnectorType(), connectorPlugin.getType(),
+                connectorContext.getCatalogName(), connectorContext.getCatalogShardName());
+            service = new ThrottlingConnectorDatabaseService(service, rateLimiter);
+        }
+
+        return service;
+    }
+
+    @Override
+    public ConnectorTableService getTableService() {
+        ConnectorTableService service = delegate.getTableService();
+
+        if (isRateLimiterEnabled()) {
```

Comment:
The throttling should be able to be enabled dynamically with FP. Could this be reworked to allow that, for example if service starts up with throttling disabled it should be possible to gradually reenable dynamically per cluster and vice versa.

Reply:
This throttle config is only to prevent the migration connector from being wrapped with the throttling decorator - this prevents double throttling first at the migration connector and then at the underlying nested connector. The MetacatRateLimiter implementation of the RateLimiter that we will have in the internal repo will be able to handle dynamic updates to throttling config changes.

Only the migrationconnector.properties will have this exempt set to true. So this config is fine as static.

Reply:
Ok makes sense